### ISSUE TITLE
lsp/_watchfiles.lua - attempt to index field 'capabilities'

### ISSUE THREAD

**Issue by woosaaahh** [neovim/neovim#23806]
### Problem

Hi everyone !

Since commit 073035a03 (May 20, 2023)], `<nvim_install_dir>/share/nvim/runtime/lua/vim/lsp/_watchFiles.lua`
will try to index `client.config.capabilities.workspace.didChangeWatchedFiles.dynamicRegistration`
which might not exists in client's configuration and so, trigger an error.

For the record, here is a list of LSP tools that don't even have the `client.config.capabilities` field : 

- Python with `jedi-language-server` + `ruff-lsp`
- Go with `gopls`
- Lua with `lua-language-server`
( all used with their latest version )

Here is an error sample when I start to edit a `*.go` file :
```
LSP[gopls]: Error SERVER_REQUEST_HANDLER_ERROR: .../opt/nvim/share/nvim/runtime/lua/vim/lsp/_watchfiles.lua:199: attempt to index field 'capabilities' (a nil value)"
```

To avoid the error for now, here is what I use in the `on_attach` function from my `lsp.lua` config :

```lua
if client.config.capabilities == nil then
  client.config.capabilities = {
    workspace = {
      didChangeWatchedFiles = {
        dynamicRegistration = true,
      },
    },
  }
end
```

### Steps to reproduce using "nvim -u minimal_init.lua"

I was able to reproduce it by just editing `minimal_init.lua` to use these lines : 
```lua
local pattern = 'go'
local cmd = {'gopls'}
```

I don't use any plugins for LSP or completion, only builtin/bundled features.

### Expected behavior

No field indexing errors.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-400+ge41b2e34b

### Language server name/version

gopls v0.11.0

### Operating system/version

Arch Linux (rolling)

### Log file

https://gist.github.com/woosaaahh/3c0d945a572c2ffa536e131f1fe35620

Thanks **a lot** to every direct/indirect contributors on the Neovim project !!! :+1:

**Comment by justinmk** [comment]
@nojnhuh

**Comment by nojnhuh** [comment]
@woosaaahh I'll add a fix for this, but I'm still not sure exactly how you're running into this error. My understanding is that the default capabilities are still being set under the hood in the case you're describing (as long as `vim.lsp.start_client` is invoked):
https://github.com/neovim/neovim/blob/700cab00680fa25e0dbdf6f91f638f4da5a09909/runtime/lua/vim/lsp.lua#L1240

Could you share the whole init file you're using? I'm not sure where minimal_init.lua is.

**Comment by woosaaahh** [comment]
Hi !

TLDR : It's fixed by ddd92a70d2aab5247895e89abaaa79c62ba7dbb4.

When I did the PR, I used Neovim up to commit e41b2e34b46a806b0fc59914263f8f4af9ac62f1.
But the first commit after, fixed this issue ddd92a70d2aab5247895e89abaaa79c62ba7dbb4.

I apologize for the troubles !
According to my PR time and ddd92a70d2aab5247895e89abaaa79c62ba7dbb4 time, it appears that I should have waited before posting my PR.

---

For the record, here are the commands I used 
```raw
$ nvim --clean -u ~/.config/neovim/minimal_init.lua
# then inside neovim
:e main.go
```

Here is the config I used to reproduce the error. I only modified line 2 & 3.
`minimal_init.lua`
```lua
--- CHANGE THESE
local pattern = 'go'
local cmd = {'gopls'}
-- Add files/folders here that indicate the root of a project
local root_markers = {'.git', '.editorconfig'}
-- Change to table with settings if required
local settings = vim.empty_dict()

vim.api.nvim_create_autocmd('FileType', {
  pattern = pattern,
  callback = function(args)
    local match = vim.fs.find(root_markers, { path = args.file, upward = true })[1]
    local root_dir = match and vim.fn.fnamemodify(match, ':p:h') or nil
    vim.lsp.start({
      name = 'bugged-ls',
      cmd = cmd,
      root_dir = root_dir,
      settings = settings
    })
  end
})
```

Here is the script I used to build and try Neovim versions
`build_neovim`
```sh
#!/bin/sh

set -e

REPO_URL='https://github.com/neovim/neovim'
REPO_DIR="$PWD/neovim"
# REPO_COMMIT='e41b2e34' # Still LSP Error
REPO_COMMIT='ddd92a70'  # No more LSP error

BUILD_TYPE="Release"
INSTALL_DIR="$HOME/opt/nvim"

# --- Clone --------------------------------------------------------------------

rm -fr -- "$REPO_DIR"
git clone "$REPO_URL" "$REPO_DIR"

cd -- "$REPO_DIR" || exit 1
git checkout "$REPO_COMMIT"

# --- Build --------------------------------------------------------------------

rm -fr -- "$INSTALL_DIR"
mkdir -p -- "$INSTALL_DIR"

make CMAKE_BUILD_TYPE="$BUILD_TYPE" CMAKE_INSTALL_PREFIX="$INSTALL_DIR"
make install
```
